tools/api_proto_breaking_change_detector/detector.py

```diff
@@ -0,0 +1,223 @@
+            #        bcode, bout, berr = run_command(f"{buf_path} mod update")
+            #        bout, berr = ''.join(bout), ''.join(berr)
+            target = Path(temp_dir, f"{Path(self._path_to_before).stem}.proto")
+
+            buf_args = [
+                "--path",
+                str(target.relative_to(
+                    Path(".").absolute())),  # buf requires relative pathing for roots...
+                "--config",
+                str(yaml_file_loc),
+            ]
+            if self._additional_args is not None:
+                buf_args.extend(self._additional_args)
+
+            copyfile(self._path_to_before, target)
+
+            initial_code, initial_out, initial_err = run_command(
+                ' '.join([buf_path, f"build -o {Path(temp_dir, BUF_STATE_FILE)}", *buf_args]))
+            initial_out, initial_err = ''.join(initial_out), ''.join(initial_err)
+
+            if initial_code != 0 or len(initial_out) > 0 or len(initial_err) > 0:
+                raise ChangeDetectorInitializeError(
+                    f"Unexpected error during init:\n\tExit Status Code: {initial_code}\n\tstdout: {initial_out}\n\t stderr: {initial_err}\n"
+                )
+
+            lock_location = Path(temp_dir, BUF_STATE_FILE)
+            with open(lock_location, "r") as f:
+                initial_lock = f.readlines()
+
+            copyfile(self._path_to_after, target)
+
+            final_code, final_out, final_err = run_command(
+                ' '.join(
+                    [buf_path, f"breaking --against {Path(temp_dir, BUF_STATE_FILE)}", *buf_args]))
+            final_out, final_err = ''.join(final_out), ''.join(final_err)
+
+            # new with buf: lock must be manually re-built
+            # but here we only re-build if there weren't any detected breaking changes
+            if len(final_out) == len(final_err) == final_code == 0:
+                _, _, _ = run_command(
+                    ' '.join([buf_path, f"build -o {Path(temp_dir, BUF_STATE_FILE)}", *buf_args]))
+            with open(lock_location, "r") as f:
+                final_lock = f.readlines()
+
+            self.initial_result = initial_code, initial_out, initial_err
+            self.final_result = final_code, final_out, final_err
+            self.initial_lock = initial_lock
+            self.final_lock = final_lock
+
+    def is_breaking(self) -> bool:
+        """Returns whether the changes between the ``before`` and ``after`` states of the proto file given in ``__init__``
+        violate any breaking change rules specified by this breaking change detector.
+
+        :return: a boolean flag indicating if the changes between ``before`` and ``after`` are breaking
+        :rtype: ``bool``
+        """
+        final_code, final_out, final_err = self.final_result
+
+        # Ways buf output could be indicative of a breaking change:
```

Comment:
This can be made more self-explanatory by having a sequence of `if` blocks, e.g.
```
if final_code != 0: return False
if final_out == "" or "Failure" in final_out: return False
if final_err == "" or "Failure" in final_err: return False
return True
```